x/pairing/keeper/pairing_subscription_test.go

```diff
@@ -0,0 +1,175 @@
+	require.Nil(t, err)
+
+	ts.ctx = testkeeper.AdvanceEpoch(ts.ctx, ts.keepers)
+
+	pairingReq := types.QueryGetPairingRequest{ChainID: ts.spec.Index, Client: consumer.Addr.String()}
+	pairing, err := ts.keepers.Pairing.GetPairing(ts.ctx, &pairingReq)
+	require.Nil(t, err)
+
+	verifyPairingQuery := &types.QueryVerifyPairingRequest{ChainID: ts.spec.Index, Client: consumer.Addr.String(), Provider: pairing.Providers[0].Address, Block: uint64(sdk.UnwrapSDKContext(ts.ctx).BlockHeight())}
+	vefiry, err := ts.keepers.Pairing.VerifyPairing(ts.ctx, verifyPairingQuery)
+	require.Nil(t, err)
+	require.True(t, vefiry.Valid)
+
+	project, err := ts.keepers.Projects.GetProjectForDeveloper(sdk.UnwrapSDKContext(ts.ctx), consumer.Addr.String(), uint64(sdk.UnwrapSDKContext(ts.ctx).BlockHeight()))
+	require.Nil(t, err)
+
+	err = ts.keepers.Projects.DeleteProject(sdk.UnwrapSDKContext(ts.ctx), project.Index)
+	require.Nil(t, err)
+
+	_, err = ts.keepers.Pairing.GetPairing(ts.ctx, &pairingReq)
+	require.NotNil(t, err)
+
+	verifyPairingQuery = &types.QueryVerifyPairingRequest{ChainID: ts.spec.Index, Client: consumer.Addr.String(), Provider: pairing.Providers[0].Address, Block: uint64(sdk.UnwrapSDKContext(ts.ctx).BlockHeight())}
+	vefiry, err = ts.keepers.Pairing.VerifyPairing(ts.ctx, verifyPairingQuery)
+	require.NotNil(t, err)
+	require.False(t, vefiry.Valid)
+}
+
+func TestRelayPaymentSubscription(t *testing.T) {
+	ts := setupForPaymentTest(t)
+	var balance int64 = 10000
+	consumer := common.CreateNewAccount(ts.ctx, *ts.keepers, balance)
+
+	_, err := ts.servers.SubscriptionServer.Subscribe(ts.ctx, &subtypes.MsgSubscribe{Creator: consumer.Addr.String(), Consumer: consumer.Addr.String(), Index: ts.plan.Index, IsYearly: false})
+	require.Nil(t, err)
+
+	ts.ctx = testkeeper.AdvanceEpoch(ts.ctx, ts.keepers)
+
+	pairingReq := types.QueryGetPairingRequest{ChainID: ts.spec.Index, Client: consumer.Addr.String()}
+	pairing, err := ts.keepers.Pairing.GetPairing(ts.ctx, &pairingReq)
+	require.Nil(t, err)
+
+	verifyPairingQuery := &types.QueryVerifyPairingRequest{ChainID: ts.spec.Index, Client: consumer.Addr.String(), Provider: pairing.Providers[0].Address, Block: uint64(sdk.UnwrapSDKContext(ts.ctx).BlockHeight())}
+	vefiry, err := ts.keepers.Pairing.VerifyPairing(ts.ctx, verifyPairingQuery)
+	require.Nil(t, err)
+	require.True(t, vefiry.Valid)
+
+	_, err = ts.keepers.Projects.GetProjectForDeveloper(sdk.UnwrapSDKContext(ts.ctx), consumer.Addr.String(), uint64(sdk.UnwrapSDKContext(ts.ctx).BlockHeight()))
+	require.Nil(t, err)
+
+	// define tests - different epoch, valid tells if the payment request should work
+	tests := []struct {
+		name  string
+		cu    uint64
+		valid bool
+	}{
+		{"happyflow", ts.spec.Apis[0].ComputeUnits, true},
+		{"epochCULimit", ts.plan.ComputeUnitsPerEpoch, false},
+	}
+
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			relayRequest := &types.RelayRequest{
+				Provider:        ts.providers[0].address.String(),
```

Comment:
this is incompatible with how we build relays after the change, and creating the object with the keys like that leaves a lof ot lines to be changed, and i am tired already of manually changing how it's done with copy paste in all of our unitests, use a common function to build stub relays please